File: src/main/java/org/retrolauncher/backend/app/platforms/infrastructure/database/memory/MemoryPlatformRepository.java
package org.retrolauncher.backend.app.platforms.infrastructure.database.memory;

import org.retrolauncher.backend.app.platforms.domain.entities.Platform;
import org.retrolauncher.backend.app.platforms.domain.repositories.PlatformRepository;

import java.util.*;

public class MemoryPlatformRepository implements PlatformRepository {
    private final Map<String, Platform> storedData = new HashMap<>();

    @Override
    public void save(Platform entity) {
        this.storedData.put(entity.getId().toString(), entity);
    }

    public void clear() {
        this.storedData.clear();
    }

    @Override
    public Optional<Platform> findById(UUID id) {
        Platform model = this.storedData.get(id.toString());
        if (model == null)
            return Optional.empty();
        return Optional.of(model);
    }

    @Override
    public List<Platform> listAll() {
        return this.storedData
                .values()
                .stream()
                .toList();
    }

    @Override
    public boolean existsByCore(String corePath) {
        return this.storedData
                .values()
                .stream()
                .anyMatch((platformModel -> platformModel.getCorePath().equals(corePath)));
    }
}

File: backend/src/app/platform/persistance/memory_platform_repository/memory_platform_repository.go
package memory_platform_repository

import platform "retrolauncher/backend/src/app/platform/domain"

type MemoryPlatformRepository struct {
	platforms map[string]*platform.Platform
}

func (r *MemoryPlatformRepository) Save(entity *platform.Platform) error {
	if r.platforms == nil {
		r.platforms = make(map[string]*platform.Platform)
	}
	r.platforms[entity.GetID().String()] = entity
	return nil
}

func (r *MemoryPlatformRepository) List() ([]*platform.Platform, error) {
	result := make([]*platform.Platform, len(r.platforms))
	i := 0

	for _, platform := range r.platforms {
		result[i] = platform
		i++
	}

	return result, nil
}

func (r *MemoryPlatformRepository) Size() int {
	if r.platforms == nil {
		return 0
	}
	return len(r.platforms)
}
